--- /d/Program Files (x86)/Microsoft Visual Studio 8/VC/PlatformSDK/include/GdiPlusimageAttributes.h	Thu Apr 14 16:54:44 2005
+++ ./GdiPlusimageAttributes.h	Mon Feb 25 17:21:58 2008
@@ -231,8 +231,8 @@
                                             nativeImageAttr,
                                             type,
                                             FALSE,
-                                            NULL,
-                                            NULL));
+                                            (ARGB)0,
+                                            (ARGB)0));
     }
 
     Status SetOutputChannel(
